art/classifiers/GPy.py

```diff
@@ -0,0 +1,254 @@
+        :type x: `np.ndarray`
+        :param y: Correct labels, one-vs-rest encoding.
+        :type y: `np.ndarray`
+        :return: Array of gradients of the same shape as `x`.
+        :rtype: `np.ndarray`
+        """
+        eps = 0.00001
+        grads = np.zeros(np.shape(x))
+        for i in range(np.shape(x)[0]):
+            # 1.0 - to mimic loss, [0,np.argmax] to get right class
+            ind = 1.0-self.predict(x[i].reshape(1, -1))[0, np.argmax(y[i])]
+            sur = 1.0-self.predict(np.repeat(x[i].reshape(1, -1), np.shape(x)[1], 0)
+                                   + eps*np.eye(np.shape(x)[1]))[:, np.argmax(y[i])]
+            grads[i] = ((sur-ind)*eps).reshape(1, -1)
+        return grads
+
+    def predict(self, x, logits=False, batch_size=128):
+        """
+        Perform prediction for a batch of inputs.
+
+        :param x: Test set.
+        :type x: `np.ndarray`
+        :param logits: `True` if the prediction should be done without squashing function.
+        :type logits: `bool`
+        :param batch_size: Size of batches. Not used in this function.
+        :type batch_size: `int`
+        :return: Array of predictions of shape `(nb_inputs, self.nb_classes)`.
+        :rtype: `np.ndarray`
+        """
+        out = np.zeros((np.shape(x)[0], 2))
+        if logits:  # output the mon-squashed version
+            out[:, 0] = self.model.predict_noiseless(x)[0].reshape(-1)
+            out[:, 1] = -1.0*out[:, 0]
+            return out
+        # output normal prediction, scale up to two values
+        out[:, 0] = self.model.predict(x)[0].reshape(-1)
+        out[:, 1] = 1.0-out[:, 0]
+        return out
+
+    def predict_uncertainty(self, x, logits=False, batch_size=128):
+        """
+        Perform prediction for a batch of inputs.
+
+        :param x: Test set.
+        :type x: `np.ndarray`
+        :param logits: `True` if the prediction should be done at the logits layer.
+        :type logits: `bool`
+        :param batch_size: Size of batches. Not used in this function.
+        :type batch_size: `int`
+        :return: Array of uncertainty predictions of shape `(nb_inputs, self.nb_classes)`.
+        :rtype: `np.ndarray`
+        """
+        # geturn uncertainty, only in one dimension
+        return self.model.predict_noiseless(x)[1]
+
+    @property
+    def layer_names(self):
+        """
+        Return the hidden layers in the model, if applicable.
+
+        :return: The hidden layers in the model, input and output layers excluded.
+        :rtype: `list`
+
+        .. warning:: `layer_names` tries to infer the internal structure of the model.
+                     This feature comes with no guarantees on the correctness of the result.
+                     The intended order of the layers tries to match their order in the model, but this is not
+                     guaranteed either.
+        """
+        raise NotImplementedError
+
+    def set_learning_phase(self, train):
```

Comment:
Remove this function because it's not required by the used classifier interfaces.